### DISCUSSION TITLE
How to upgrade dify in a good way?

### DISCUSSION THREAD

**Discussion by seikyo-cho-lvgs** [langgenius/dify#8154] [General]
### Self Checks

- [X] I have searched for existing issues [search for existing issues](https://github.com/langgenius/dify/issues), including closed ones.
- [X] I confirm that I am using English to submit this report (我已阅读并同意 [Language Policy](https://github.com/langgenius/dify/issues/1542)).
- [X] [FOR CHINESE USERS] 请务必使用英文提交 Issue，否则会被关闭。谢谢！:）
- [X] Please do not modify this template :) and fill in all the required fields.

### Content

Dify releases new versions very frequently. If deploy dify on kubernetes by helm chart, is there a good way to upgrade?

The common way is set new version's docker image tag to each service's image setting, then apply. Is it okay to always pull the newest tag by cronjob and deploy to production environment directly?

Something like API, frontend, and database migration, should be done together. If insure the database migration run first, is it safe to apply other services automatically?

**Comment by CXwudi** [comment]
I used to pulled the latest tag through watchtower. However, I realized the upgrade process through watchtower will break the vector DB. 

Specifically, after the upgrade through watchtower, I go to the retrieval testing shown in the screenshot below:

![image](https://github.com/user-attachments/assets/3bb034d3-8d3a-4861-8bb0-7b2feba36e39)

it will failed with:

```
Error during query: [{'locations': [{'column': 6, 'line': 1}], 'message': 'Cannot query field "Vector_index_e84bdc53_76ed_493c_b9b9_0ad03baa4bd1_Node" on type "GetObjectsObj".', 'path': None}]; 
Error during query: [{'locations': [{'column': 6, 'line': 1}], 'message': 'Cannot query field "Vector_index_e84bdc53_76ed_493c_b9b9_0ad03baa4bd1_Node" on type "GetObjectsObj".', 'path': None}]
```

Hence I realized that you can't simply just pull the latest version. I guess maybe I should stop the watchtower? but IDK

**Reply by seikyo-cho-lvgs** [reply]
Thank you very much for the very useful information!
Then it's better test in staging environment first, if everything works well we can upgrade in production environment for each new release.

**Comment by CXwudi** [comment]
Hi, I confirmed that watchtower is the causes that break the vector DB as I mentioned in https://github.com/langgenius/dify/discussions/8154#discussioncomment-10595906, probably because watchtower won't upgrade `api`, `worker`, and `web` all at the same time and hence the mismatched version could be cause. With that being said, it is safe to just pull the newest tag and restart dify, assuming you restart `api`, `worker`, and `web` all at the same time, and also prevent watchtower from touching these 3 containers.

Here is my script that just worked yesterday that updated my Dify from 0.8.2 to 0.8.3:

```sh
#!/bin/bash

BASE_COMPOSE_FILE="compose.base.yml"
PROD_COMPOSE_FILE="compose.prod.yml"
SERVICES="api worker web"

# Check if compose files exist
if [ ! -f "$BASE_COMPOSE_FILE" ] || [ ! -f "$PROD_COMPOSE_FILE" ]; then
    echo "One or both Compose files not found: $BASE_COMPOSE_FILE, $PROD_COMPOSE_FILE"
    exit 1
fi

# Function to run docker compose commands with both files
docker_compose() {
    docker compose -f "$BASE_COMPOSE_FILE" -f "$PROD_COMPOSE_FILE" "$@"
}

# Check for updates
echo "Checking for updates..."
update_available=false
output=$(docker_compose pull --quiet 2>&1)

# Check if there was an error
if [ $? -ne 0 ]; then
    echo "An error occurred:"
    echo "$output"
    exit 1
fi

# Process each line of output
while IFS= read -r line; do
    if [[ "$line" == *"Image is up to date"* ]]; then
        service=$(echo "$line" | awk '{print $1}')
        echo "$service: Up to date"
    elif [[ "$line" == *"Pulling"* ]]; then
        service=$(echo "$line" | awk '{print $2}')
        echo "$service: Update available and pulled"
        update_available=true
    fi
done <<< "$output"

# If no updates are available, exit
if [ "$update_available" = false ]; then
    echo "No updates available. Exiting."
    exit 0
fi

# If updates are available, proceed with restart
echo "Updates available. Proceeding with restart..."

# Stop and remove containers for all services
echo "Stopping and removing containers for all services..."
docker_compose rm -sf $SERVICES

# Start all services with new images
echo "Starting all services with new images..."
docker_compose up -d $SERVICES

# Prune dangling images
echo "Pruning dangling images..."
docker image prune -f

echo "All services ($SERVICES) have been updated to their latest versions and dangling images have been removed"
```

Noted that I have two docker compose file together created a production deployment
